zerver/tests/test_message_edit.py

```diff
@@ -1655,3 +1657,114 @@ def test_user_group_mention_restrictions_while_editing(self) -> None:
             },
         )
         self.assert_json_success(result)
+
+    def test_move_message_to_new_topic_with_automatic_follow_policy(self) -> None:
+        self.login("iago")
+        hamlet = self.example_user("hamlet")
+        shiva = self.example_user("shiva")
+        stream = self.make_stream("new_stream")
+        self.subscribe(hamlet, stream.name)
+        self.subscribe(shiva, stream.name)
+
+        self.send_stream_message(
+            sender=hamlet,
+            stream_name="new_stream",
+            topic_name="test",
+            content="Never here",
+        )
+        msg_id = self.send_stream_message(
+            sender=hamlet,
+            stream_name="new_stream",
+            topic_name="test",
+            content="Always here",
+        )
```

Comment:
A few comments on the test:

- Let's use more semantic names for your topics. `test` and `testing` don't tell the reader anything about the intent. Names like `original` and `post_move` would be better, though see below, because you're likely to . You could also think about just having a variable with a name like `original_topic` rather than being expressive with the string topic names.
- I don't understand why the `Shiva` user gets referenced at all. Was the intent to have them move topics, via `self.api_patch` so you don't have to switch who is logged in, but you later realized Hamlet had the permission? Though maybe you do need this user, see the next note.
- This test doesn't verify the full set of logic/conditions for this code. It's important to verify both that (1) a user whose message is moved to be the first message in a new topic has their visibility policy changed but also (2) a user who sent a message that is moved to be a not-first message in a new topic, or into an existing topic, does not have their visibility policy changed. Your tests don't attempt to verify the various corner cases around that -- like what happens if you move an older message to be the new first message in an existing topic? The tests should verify that whatever happens in that case is as intended.

Likely to address this you'll need to extend these tests to do more setup and/or make multiple moves. So, for example, you could have:
- Iago sends a message to "original"
- Cordelia sends a message to "original"
- Hamlet sends a message to "original"
- Shiva sends a message to "original"
- Confirm only Iago follows the topic "original".
- Move Hamlet&Shiva messages to "post-move"
- Confirm Hamlet now follows "post-move"
- Move Cordelia message to "post-move", which notably is now the first message in the topic.
- Confirm cordelia does now follow "post-move" too.

Reply:
You probably also need some form of test coverage for the private stream with protected history case.

Reply:
updated!